apps/prairielearn/src/pages/studentInstanceQuestionHomework/studentInstanceQuestionHomework.js

```diff
@@ -86,59 +88,110 @@ function processSubmission(req, res, callback) {
   );
 }
 
-router.post('/', function (req, res, next) {
-  if (res.locals.assessment.type !== 'Homework') return next();
+router.post(
+  '/',
+  asyncHandler(async function (req, res, next) {
+    if (res.locals.assessment.type !== 'Homework') return next();
 
-  if (!res.locals.authz_result.authorized_edit) {
-    return next(error.make(403, 'Not authorized', res.locals));
-  }
+    if (!res.locals.authz_result.authorized_edit) {
+      return next(error.make(403, 'Not authorized', res.locals));
+    }
 
-  if (req.body.__action === 'grade' || req.body.__action === 'save') {
-    processSubmission(req, res, function (err, variant_id) {
-      if (ERR(err, next)) return;
+    if (req.body.__action === 'grade' || req.body.__action === 'save') {
+      if (res.locals.assessment.group_work) {
+        const groupConfig = await groupAssessmentHelper.getGroupConfig(res.locals.assessment.id);
+        if (groupConfig.has_roles) {
+          const result = await groupAssessmentHelper.getQuestionPermissions(
+            res.locals.assessment_question.id,
+            res.locals.user.user_id,
+          );
+          // Users without the correct roles cannot submit
+          if (!result.can_submit) {
+            return next(
+              error.make(
+                403,
+                'Current group roles have no permission to submit this question',
+                res.locals,
+              ),
+            );
+          }
+
+          // Get group information
+          const groupId = await groupAssessmentHelper.getGroupId(
+            res.locals.assessment.id,
+            res.locals.user.user_id,
+          );
+          const groupInfo = await groupAssessmentHelper.getGroupInfo(groupId, groupConfig);
+          const { validationErrors, rolesAreBalanced, usersWithoutRoles } = groupInfo.rolesInfo;
+
+          // Prevent submissions if role configuration is invalid
```

Comment:
Would it make more sense to do the validation first, and only complain about missing permissions if the config is valid?

Reply:
I think that order makes sense, checking for an invalid role configuration before individual permissions. Both of these were safeguards in case a user crafts a request to submit when the UI shouldn't allow them to anyways, so hopefully neither should be too common.